src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/TypeFixupSignature.cs

```diff
@@ -39,11 +42,114 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
             }
 
             return dataBuilder.ToObjectData();
         }
 
+        private static void EncodeTypeLayout(ObjectDataSignatureBuilder dataBuilder, TypeDesc type)
+        {
+            Debug.Assert(type.IsValueType);
+            MetadataType defType = (MetadataType)type;
+
+            int pointerSize = type.Context.Target.PointerSize;
+            int size = defType.InstanceFieldSize.AsInt;
+            int alignment = GetClassAlignmentRequirement(defType);
+            ReadyToRunTypeLayoutFlags flags = ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_Alignment | ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_GCLayout;
+            if (alignment == pointerSize)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_Alignment_Native;
+            }
+
+            if (!defType.ContainsGCPointers)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_GCLayout_Empty;
+            }
+
+            bool isHfa = defType.IsHfa && defType.Context.Target.Architecture == TargetArchitecture.ARM || defType.Context.Target.Architecture == TargetArchitecture.ARM64;
+            if (isHfa)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_HFA;
+            }
+
+            dataBuilder.EmitUInt((uint)flags);
+            dataBuilder.EmitUInt((uint)size);
+
+            if (isHfa)
+            {
+                switch (defType.HfaElementType.Category)
+                {
+                    case TypeFlags.Single:
+                        dataBuilder.EmitUInt((uint)CorElementType.ELEMENT_TYPE_R4);
+                        break;
+                    case TypeFlags.Double:
+                        dataBuilder.EmitUInt((uint)CorElementType.ELEMENT_TYPE_R4);
+                        break;
+                }
+            }
+            
+            if (alignment != pointerSize)
+            {
+                dataBuilder.EmitUInt((uint)alignment);
+            }
+
+            if (defType.ContainsGCPointers)
+            {
+                // Encode the GC pointer map
+                GCPointerMap gcMap = GCPointerMap.FromInstanceLayout(defType);
+
+                byte[] encodedGCRefMap = new byte[(size / pointerSize + 7) / 8];
+                int bitIndex = 0;
+                foreach (bool bit in gcMap)
+                {
+                    if (bit)
+                    {
+                        encodedGCRefMap[bitIndex / 8] += (byte)(1 << (bitIndex & 7));
+                    }
+
+                    ++bitIndex;
+                }
+
+                dataBuilder.EmitBytes(encodedGCRefMap);
+            }
+        }
+
+        /// <summary>
+        /// Managed implementation of CEEInfo::getClassAlignmentRequirementStatic
+        /// </summary>
+        private static int GetClassAlignmentRequirement(MetadataType type)
+        {
+            int alignment = type.Context.Target.PointerSize;
+
+            if (((MetadataType)type).HasLayout())
```

Comment:
Nit - the typecast seems superfluous, the method argument is already a <code>MetadataType</code>.

Reply:
Fixed